[UnifiedPDF] "Copy Link" context menu item should copy the link from the current selection

aprotyas opened the pull request:
#### 2a14de2051cd21485f7d7a65f8caf7b9f3d9ffec
<pre>
[UnifiedPDF] &quot;Copy Link&quot; context menu item should copy the link from the current selection
<a href="https://bugs.webkit.org/show_bug.cgi?id=269423">https://bugs.webkit.org/show_bug.cgi?id=269423</a>
<a href="https://rdar.apple.com/122981243">rdar://122981243</a>

Reviewed by Simon Fraser.

In 274611@main, we added support for a &quot;Copy Link&quot; context menu item.
This patch imbibes the item with its namesake functionality through the
performCopyLinkOperation method. Said method checks if the context menu
was invoked over a link-bearing annotation, and if so, extracts the URL
from that annotation and puts it on the pasteboard.

Note that we introduce a subtle behavior change through this patch. We
no longer unconditionally present the &quot;Copy Link&quot; context menu item for
any link annotations, and only do so for annotations containing external
links (not destinations).

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::convertFromDocumentToPlugin const):
(WebKit::UnifiedPDFPlugin::convertFromPageToDocument const):
(WebKit::annotationIsExternalLink):
(WebKit::annotationIsLinkWithDestination):
(WebKit::UnifiedPDFPlugin::handleContextMenuEvent):
(WebKit::UnifiedPDFPlugin::createContextMenu const):
(WebKit::UnifiedPDFPlugin::selectionContextMenuItems const):
(WebKit::UnifiedPDFPlugin::performContextMenuAction):
(WebKit::UnifiedPDFPlugin::performCopyLinkOperation const):

Canonical link: <a href="https://commits.webkit.org/274695@main">https://commits.webkit.org/274695@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8e6e24fe49a6d1813757c1a08554d2ab0f8841a6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39777 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18788 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42151 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42321 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/35679 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21679 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16118 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40351 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15823 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/34402 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/13703 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13719 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43600 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36115 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35689 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/39465 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/14592 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12006 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37746 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16211 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8918 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16259 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/15855 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
```diff
@@ -338,16 +339,14 @@ class UnifiedPDFPlugin final : public PDFPluginBase, public WebCore::GraphicsLay
     WebCore::IntPoint convertFromDocumentToPlugin(const WebCore::IntPoint&) const;
-
+    WebCore::IntRect convertFromDocumentToPlugin(const WebCore::IntRect&) const;
+    std::optional<PDFDocumentLayout::PageIndex> pageIndexForDocumentPoint(const WebCore::IntPoint&) const;
```
I prefer this down where it was, outside this grouping of coordinate conversion functions.

aprotyas replied:
Okay, I'll move it back.

aprotyas replied:
https://github.com/WebKit/WebKit/commit/8e6e24fe49a6d1813757c1a08554d2ab0f8841a6

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1828,6 +1850,32 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
+    if (!annotationIsExternalLink(annotation.get()))
+        return;
+
+    NSURL *url = [annotation URL];
```
`RetainPtr url =`

aprotyas replied:
https://github.com/WebKit/WebKit/commit/8e6e24fe49a6d1813757c1a08554d2ab0f8841a6

aprotyas committed:
[UnifiedPDF] "Copy Link" context menu item should copy the link from the current selection
https://bugs.webkit.org/show_bug.cgi?id=269423
rdar://122981243

Reviewed by Simon Fraser.

In 274611@main, we added support for a "Copy Link" context menu item.
This patch imbibes the item with its namesake functionality through the
performCopyLinkOperation method. Said method checks if the context menu
was invoked over a link-bearing annotation, and if so, extracts the URL
from that annotation and puts it on the pasteboard.

Note that we introduce a subtle behavior change through this patch. We
no longer unconditionally present the "Copy Link" context menu item for
any link annotations, and only do so for annotations containing external
links (not destinations).

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::convertFromDocumentToPlugin const):
(WebKit::UnifiedPDFPlugin::convertFromPageToDocument const):
(WebKit::annotationIsExternalLink):
(WebKit::annotationIsLinkWithDestination):
(WebKit::UnifiedPDFPlugin::handleContextMenuEvent):
(WebKit::UnifiedPDFPlugin::createContextMenu const):
(WebKit::UnifiedPDFPlugin::selectionContextMenuItems const):
(WebKit::UnifiedPDFPlugin::performContextMenuAction):
(WebKit::UnifiedPDFPlugin::performCopyLinkOperation const):

Canonical link: https://commits.webkit.org/274695@main
